_posts/2024-08-27-building-browser.md

```diff
@@ -0,0 +1,175 @@
+    alt="Verso displaying two windows"></a>
+<figcaption>Figure 2: Verso creates two different windows with same OpenGL context.</figcaption></figure>
+
+There is still room for improvement.
+For example, WebRender currently [only supports rendering a single “document”](https://github.com/servo/webrender/blob/0.64/webrender/src/renderer/mod.rs#L1228).
+This means this document has to constantly update different display lists to show on other windows, which may cause race conditions and let unintended webview display on the wrong window in a slip second.
+<!-- why is this not a problem in Firefox? -->
+
+There are also different OpenGL versions across multiple platforms, which can be challenging to configure and link.
+Verso is experimenting with using [Glutin](https://docs.rs/glutin/latest/glutin/) for better configuration and attempting to get closer to the general Rust ecosystem.
+
+## What’s next?
+
+With multi-view and multi-window support as the fundamental building blocks, we could create more UI elements to keep pushing the envelope of our browser and embedding research.
+At the same time, Servo is a huge project, with many potential improvements still to come, so we want to reflect on our progress and decide on our priorities.
+Here are some directions that are worth pursuing.
+
+### Are we browser yet?
+
+We are building a page, [arewebrowseryet.com](https://servo.zulipchat.com/#narrow/stream/263398-general/topic/arewebrowseryet.2Ecom), in the hope of gathering the strength of the community to help us display the statistics of supported CSS properties and web APIs in Servo by [popularity order](https://chromestatus.com/metrics/css/popularity) and benchmark results such as jetstream2 and speedometer3.
+We hope this will eventually give newcomers a better overview of Servo.
+
+### Script triage
+
+There’s a [Servo triage meeting](https://github.com/servo/project/issues/100) every two weeks to triage any issues around the script crate and more.
+Once we get the statistics of supported web APIs, we can find the most popular ones that haven’t been implemented or fixed yet.
+We are already fixing some issues around loading the order and re-implementing [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream) in Rust.
+If you are interested in implementing web APIs in Servo, feel free to join the next meeting.
+
+### Multi-process and sandboxing
+
+Some features are crucial to the browser but not visible to users.
+Multi-process architecture and sandboxing belong to this category.
+Both of these are implemented in Servo to some extent, but are not enabled by default <!-- is this true for both? i know it’s true for multiprocess, but what about sandboxing? -->.
+They also only support Linux and macOS for now <!-- is this true for both? i know it’s true for sandboxing, but what about multiprocess? -->.
```

Comment:
ipc-channel supports Windows, but there's [a bug that will crash on Windows](https://github.com/servo/ipc-channel/issues/310) if we try to enable multiprocess mode.
I belive both mods only support Linux and macOS for now in [constellation/sandboxing.rs](https://github.com/servo/servo/blob/ebed9218f2907767ba3c9dd9f27f30a6a6e9f225/components/constellation/sandboxing.rs).
And the [default Opts](https://doc.servo.org/src/servo_config/opts.rs.html#396) disable them by default.